### ISSUE TITLE
Entries in Context array in wrong order

### ISSUE THREAD

**Issue by lauramartingonzalezzz** [ScorpioBroker/ScorpioBroker#480]
Hello, 
we have encountered a problem regarding the order in the _context_ property. Since some recent version (I think it is v4.0.6), the Scorpio self-hosts the core context, in this case, NGSI-LD v1.6.1 (currently it is v1.7.1). When requesting entities with a custom Link header, the response looks like this:
```
Request:
GET /ngsi-ld/v1/entities/?type=Catalogue
Accept: application/ld+json
Link: <https://raw.githubusercontent.com/smart-data-models/dataModel.DCAT-AP/master/context.jsonld>;rel="http://www.w3.org/ns/json-ld#context";type="application/ld+json"

Response:
{
  "id": "urn:ngsi-ld:Catalogue:test1",
  "type": "Catalogue",
  ...
  "@context": [
      "https://raw.githubusercontent.com/smart-data-models/dataModel.DCAT-AP/master/context.jsonld",
      "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context-v1.7.jsonld"
  ]
}
```

As you can see, the ngsi-ld-core context appers in second place. 
Here's the issue: If the entity has a property whose name appears in both contexts (or in more than one in case of multiple contexts), the namespace selected is the last one, i.e., the last context appearing in the _context_ property. For example, if we are using a custom context with a property named _dataset_ it would be overriden by the core context (it also has a property named _dataset_). This makes it impossible to use custom attributes if they appear in the core context. We think it would make more sense if the core context went first because of these kinds of situations.
So... is it working as expected or should the order be reversed?

Thanks!
Laura

**Comment by amitrnec** [comment]
Hi,
It is working as expected.
As the Core @context is protected and cannot be overridden, when performing term to URI expansion or compaction, implementations shall always consider the Core @context as having absolute precedence, regardless of the position of the Core @context in the @context array of elements. Nonetheless, NGSI-LD data providers may use appropriate term prefixing to ensure that a proper term to URI expansion or compaction is performed.

**Comment by lauramartingonzalezzz** [comment]
ok, I guess we will find a workaround for our platform
thanks for the info!

**Comment by jason-fox** [comment]
@lauramartingonzalezzz  Here is a valid JSON-LD workaround:  http://tinyurl.com/2s4jvf6v

Imagine I want have a property with the short name `createdAt`, I cannot overwrite the existing non-reified property defined in the core `@context` so I am forced to prefix my user context entry. This is written in the NGSI-LD specification

> *4.6.2 Supported names*
> Even though the JSON serialization format allows inclusion of any character in the Unicode space, NGSI-LD restricts
> Entity Type Names, Property Names and Relationship Names to the following ABNF grammar... 
> In order to avoid name clashing, names can be prefixed as specified by the following BNF grammar:
> 
> `prefix = unicodeLetter *nameChar`
> `name =/ prefix %x3A unicodeLetter *nameChar ; prefix:name 

Now `%x3A` is just a `:`, so the following is a valid expandable JSON-LD payload for NGSI-LD:


```json
{
  "id": "urn:001",
  "createdAt": "2012-12-21 ...",
  "ex:createdAt": {
    "value": "2012-12-21 ...",
    "type": "Property"
  },
  "type": "Something",
  "@context": [
    "https://uri.etsi.org/ngsi-ld/v1/ngsi-ld-core-context-v1.7.jsonld",
    {
      "ex": "http://example.com/",
      "ex:expiresAt": "ex:expiresAt"
    }
  ]
}
```

In reality, **neither** attribute is just `createdAt` on its own, they are both separate URIs. If you really need to map your own `http://example.com/createdAt` back to `createdAt`, you can make a separate expansion/compaction operation to switch. The resultant payload will no longer be valid NGSI-LD, but will be valid JSON-LD for internal use. Do the reverse when pushing back to the context broker.
 
In summary, NGSI-LD core terms can never be redefined to something else if you expect the payload to be NSGI-LD.

**Comment by lauramartingonzalezzz** [comment]
thanks!! we will try it out
these kind of things are quite tricky when you try to be compliant with Smart Data Models too, but this is good to know
thanks again!
